contrib/mesos/pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,923 @@
+	"github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/scheduler/meta"
+	"github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/scheduler/metrics"
+	"github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/scheduler/podtask"
+	"github.com/GoogleCloudPlatform/kubernetes/contrib/mesos/pkg/scheduler/uid"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet/container"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/tools"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	log "github.com/golang/glog"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+	mutil "github.com/mesos/mesos-go/mesosutil"
+	bindings "github.com/mesos/mesos-go/scheduler"
+)
+
+type Slave struct {
+	HostName string
+}
+
+func newSlave(hostName string) *Slave {
+	return &Slave{
+		HostName: hostName,
+	}
+}
+
+type slaveStorage struct {
+	sync.Mutex
+	slaves map[string]*Slave // SlaveID => slave.
+}
+
+func newSlaveStorage() *slaveStorage {
+	return &slaveStorage{
+		slaves: make(map[string]*Slave),
+	}
+}
+
+// Create a mapping between a slaveID and slave if not existing.
+func (self *slaveStorage) checkAndAdd(slaveId, slaveHostname string) {
+	self.Lock()
+	defer self.Unlock()
+	_, exists := self.slaves[slaveId]
+	if !exists {
+		self.slaves[slaveId] = newSlave(slaveHostname)
+	}
+}
+
+func (self *slaveStorage) getSlaveIds() []string {
+	self.Lock()
+	defer self.Unlock()
+	slaveIds := make([]string, 0, len(self.slaves))
+	for slaveID := range self.slaves {
+		slaveIds = append(slaveIds, slaveID)
+	}
+	return slaveIds
+}
+
+func (self *slaveStorage) getSlave(slaveId string) (*Slave, bool) {
+	self.Lock()
+	defer self.Unlock()
+	slave, exists := self.slaves[slaveId]
+	return slave, exists
+}
+
+type PluginInterface interface {
+	// the apiserver may have a different state for the pod than we do
+	// so reconcile our records, but only for this one pod
+	reconcilePod(api.Pod)
+
+	// execute the Scheduling plugin, should start a go routine and return immediately
+	Run(<-chan struct{})
+}
+
+// KubernetesScheduler implements:
+// 1: A mesos scheduler.
+// 2: A kubernetes scheduler plugin.
+// 3: A kubernetes pod.Registry.
+type KubernetesScheduler struct {
+	// We use a lock here to avoid races
+	// between invoking the mesos callback
+	// and the invoking the pod registry interfaces.
+	// In particular, changes to podtask.T objects are currently guarded by this lock.
+	*sync.RWMutex
+
+	// Config related, write-once
+
+	schedcfg          *schedcfg.Config
+	executor          *mesos.ExecutorInfo
+	executorGroup     uint64
+	scheduleFunc      PodScheduleFunc
+	client            *client.Client
+	etcdClient        tools.EtcdGetSet
```

Comment:
IIUC the scheduler is using etcd directly for two things: (1) master election, (2) storing framework ID? If so, please add a comment about that here. (If I missed some usage, please add it too, of course.)

In theory, only the API server is supposed to talk to etcd. Our plan is for the API server to provide a master election service (see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/proposals/high-availability.md), which addresses (1), and to have a config store that any component can use (see #6923), which addresses (2). So what you're doing here isn't quite kosher. That said, I think we've already violated this principle in #9133, and I don't see any other way for you to do what you're doing. But please add a TODO saying to move usage (1) to the API-server-based master election service when it is written, and (2) to the config store when it is written.